plaso/parsers/aws_elb_access.py

```diff
@@ -0,0 +1,312 @@
+# -*- coding: utf-8 -*-
+"""Parser for AWS ELB access logs.
+
+This parser is based on the log format documented at
+https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-access-logs.html
+"""
+import pyparsing
+
+from dfdatetime import time_elements as dfdatetime_time_elements
+
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import definitions
+from plaso.lib import errors
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class AWSELBEventData(events.EventData):
+  """AWS Elastic Load Balancer access log event data
+
+  Attributes:
+    type (str): The type of request or connection.
+    elb (str): The resource ID of the load balancer.
+    client_ip_address (str): The IP address of the requesting client.
+    client_port (int): The port of the requesting client.
+    target_ip_address (str): The IP address of the target that processed
+        this request.
+    target_port (int): The port of the target that processed this request.
+    request_processing_time (str): The total time elapsed from
+        the time the load balancer received the request until the
+        time it sent the request to a target.
+    target_processing_time (str): The total time elapsed from
+        the time the load balancer sent the request to a target until
+        the target started to send the response headers.
+    response_processing_time (str): The total processing time elapsed.
+    target_status_code (int): The status code of the response from the target.
+    received_bytes (int): The size of the request, in bytes, received from
+        the client.
+    sent_bytes (int): The size of the response, in bytes, sent to the client.
+    user_agent (str): A User-Agent string.
+    ssl_cipher (str): [HTTPS listener] The SSL cipher.
+    ssl_protocol (str): [HTTPS listener] The SSL protocol.
+    target_group_arn (str): The Amazon Resource Name (ARN) of the target group.
+    trace_id (str): The contents of the X-Amzn-Trace-Id header.
+    domain_name (str): [HTTPS listener] The SNI domain provided by the
+        client during the TLS handshake.
+    chosen_cert_arn (str): [HTTPS listener] The ARN of the certificate
+        presented to the client.
+    matched_rule_priority (int): The priority value of the rule that
+        matched the request.
+    request_creation_time (str): The time when the load balancer received the
```

Comment:
why is this an attribute and not an event?

Reply:
since the log entries are generated by the load balancer, we want the event time to be the "time" field from the log, which is the time when the load balancer generated a response to the client. If we use the request_creation_time for the event timestamp it may create some confusion since it deviates from how AWS interprets the time fields.

Reply:
why not create a separate request event ?

Reply:
Two events are now generated, each with their own timestamp and description. One event's timestamp uses the time the load balancer received the request from the client as it's timestamp, and the other event uses the time the load balancer generated a response to the client.